pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,346 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/sethvargo/go-envconfig"
+
+	"github.com/abcxyz/pkg/testutil"
+)
+
+const (
+	testAppID     = "asdf"
+	testVersion   = "1.0.0"
+	testInstallID = "yv66vt6tvu8="
+	testServerURL = "https://example.com"
+)
+
+func defaultClient() Client {
+	return Client{
+		appID:      testAppID,
+		version:    testVersion,
+		installID:  testInstallID,
+		httpClient: &http.Client{Timeout: 1 * time.Second},
+		optOut:     false,
+		config: &metricsConfig{
+			ServerURL: testServerURL,
+			NoMetrics: false,
+		},
+	}
+}
+
+// Not all failure cases can be easily tested, will test subset that is easy
+// to reproduce.
+func Test_New_unhappy(t *testing.T) {
+	t.Parallel()
+	cases := []struct {
+		name      string
+		appID     string
+		env       map[string]string
+		wantError string
+	}{
+		{
+			name:      "empty_app_id_noop",
+			appID:     "",
+			wantError: "appID cannot be empty",
+		},
+		{
+			name:      "opt_out_env_noop_no_err",
+			appID:     testAppID,
+			env:       map[string]string{"NO_METRICS": "TRUE"},
+			wantError: "",
+		},
+		{
+			name:      "bad_url_noop",
+			appID:     testAppID,
+			env:       map[string]string{"METRICS_URL": "htttpq://%foo*(*fg.com4/\\"},
+			wantError: "failed to parse server URL",
+		},
+	}
+	for _, tc := range cases {
+		tc := tc
```

Comment:
```suggestion
	}

	for _, tc := range cases {
```

The shadow isn't required in Go 1.22+